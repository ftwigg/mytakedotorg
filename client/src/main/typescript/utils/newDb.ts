export const documents = [
  {
    title: "United States Constitution",
    primaryDate: "1788-06-21",
    primaryDateKind: "ratified",
    kind: "document",
    url: "gD78kfjJVp5yKdVVDeDeZL03ViA4UacyOvXZFzHnQaY="
  },
  {
    title: "Bill of Rights",
    primaryDate: "1791-12-15",
    primaryDateKind: "ratified",
    kind: "document",
    url: "uvzrf9zoJssiqWJw7jPvk8O_o3uQcO4WV-zyVD6EQlE="
  },
  {
    title: "Amendment 11",
    primaryDate: "1795-02-07",
    primaryDateKind: "ratified",
    kind: "document",
    url: "5NYBgn8DlGfSS3UxSpe4_BxmpFlnS84DOwoL9ZaB7k4="
  },
  {
    title: "Amendment 12",
    primaryDate: "1804-06-15",
    primaryDateKind: "ratified",
    kind: "document",
    url: "HHLcwaOEUWBdiww5uqEVBAQ9bGA9bV94cQbg5pBEonc="
  },
  {
    title: "Amendment 13",
    primaryDate: "1865-12-06",
    primaryDateKind: "ratified",
    kind: "document",
    url: "anw0aNyzDzXJIMFc3Pdx21KLwjFWS-xe1PKT7sdT5tY="
  },
  {
    title: "Amendment 14",
    primaryDate: "1868-07-09",
    primaryDateKind: "ratified",
    kind: "document",
    url: "EoKRQLGeFGCfoeRPqPVuFhaMj1QLbnfWfHmzhUGNwLU="
  },
  {
    title: "Amendment 15",
    primaryDate: "1870-02-03",
    primaryDateKind: "ratified",
    kind: "document",
    url: "unF7vHwSF3ud3AJv_LE1CVx4oq28hWsfJAJROVPnAeE="
  },
  {
    title: "Amendment 16",
    primaryDate: "1913-02-03",
    primaryDateKind: "ratified",
    kind: "document",
    url: "LUySG57K1-SvME4eCIgqOifXtIXEi99vyiNBvx2zmXU="
  },
  {
    title: "Amendment 17",
    primaryDate: "1913-04-08",
    primaryDateKind: "ratified",
    kind: "document",
    url: "60RladvsWpRRxbcYymuRdgVzdZKYU2iKS7aSZgL-orI="
  },
  {
    title: "Amendment 18",
    primaryDate: "1919-01-16",
    primaryDateKind: "ratified",
    kind: "document",
    url: "LoEk_oxGIMLvAR8Fzo_bdSvLpMTT6yedeydxkheH3Qk="
  },
  {
    title: "Amendment 19",
    primaryDate: "1920-08-18",
    primaryDateKind: "ratified",
    kind: "document",
    url: "G8mjqCWAoAShguyukLP5XkVFnZ-wV5EWSTjKRP5iuXQ="
  },
  {
    title: "Amendment 20",
    primaryDate: "1933-01-23",
    primaryDateKind: "ratified",
    kind: "document",
    url: "IrdYwJgL2luB_2YL3H7XZREuF-zMwbquniTsgDktf0I="
  },
  {
    title: "Amendment 21",
    primaryDate: "1933-12-05",
    primaryDateKind: "ratified",
    kind: "document",
    url: "VVvhifhLz3EkpAyB8Eig_d7vjcrQBrbQtMuXhsGG9y4="
  },
  {
    title: "Amendment 22",
    primaryDate: "1951-02-27",
    primaryDateKind: "ratified",
    kind: "document",
    url: "9YXRCpeu3iHLfhp2QhkDJCUhXLvTioGMW2HU5icJiHY="
  },
  {
    title: "Amendment 23",
    primaryDate: "1961-03-29",
    primaryDateKind: "ratified",
    kind: "document",
    url: "P1jNjqVqu5df7AV53bztVSGVZ73X0GuL0bGn-r5RAK4="
  },
  {
    title: "Amendment 24",
    primaryDate: "1964-01-23",
    primaryDateKind: "ratified",
    kind: "document",
    url: "oUX4LOq2wrfbM5NvVcndK6223PvjdAh4MwMlawlaeq8="
  },
  {
    title: "Amendment 25",
    primaryDate: "1967-02-10",
    primaryDateKind: "ratified",
    kind: "document",
    url: "re_S315SZuz7D0d2PybH3TUj7a9pGMa5AT96Lz_2a7E="
  },
  {
    title: "Amendment 26",
    primaryDate: "1971-07-01",
    primaryDateKind: "ratified",
    kind: "document",
    url: "-188YXOj6s7ulIWDEqTtYwmifFR30eD3inVrky9HcP4="
  },
  {
    title: "Amendment 27",
    primaryDate: "1992-05-05",
    primaryDateKind: "ratified",
    kind: "document",
    url: "a4rvlRzn6OjfY_pWeVroPGP4oMWFDSBQwwXQh1U3pus="
  }
];
