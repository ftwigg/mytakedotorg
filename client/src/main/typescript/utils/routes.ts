export const routes = {
  DRAFTS: "/drafts/",
  DRAFTS_NEW: "/drafts/new",
  DRAFTS_DELETE: "/drafts/delete",
  DRAFTS_PUBLISH: "/drafts/publish",
  DRAFTS_SAVE: "/drafts/save",
  LOGIN: "/login",
  LOGOUT: "/logout"
};
