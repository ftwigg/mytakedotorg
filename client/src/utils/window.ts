/**
* Small screen navigation toggle
**/
function addNavEvents() {
  let body = document.body;
  let navToggle = document.getElementsByClassName("header__nav-toggle")[0];
  let navMenu = document.getElementsByClassName("nav")[0];

  function navEvent(e: Event) {
    if (!e.defaultPrevented) {
      toggleNav();
      e.preventDefault();
    }
  }

  function bodyEvent(e: Event) {
    if (body.classList.contains("fade") && !e.defaultPrevented) {
      if (
        e.srcElement &&
        !e.srcElement.classList.contains("header__nav--icon")
      ) {
        if (
          !e.srcElement.classList.contains("nav__link") &&
          !e.srcElement.classList.contains("nav__link-text")
        ) {
          toggleNav();
          e.preventDefault();
        }
      }
    }
  }

  if (navToggle) {
    navToggle.addEventListener("click", navEvent);
  }

  body.addEventListener("click", bodyEvent);
  body.addEventListener("touchstart", bodyEvent);

  function toggleNav() {
    if (navMenu) {
      if (navMenu.classList.contains("collapse")) {
        navMenu.classList.remove("collapse");
      } else {
        navMenu.classList.add("collapse");
      }
    }

    if (body.classList.contains("fade")) {
      body.classList.remove("fade");
    } else {
      body.classList.add("fade");
    }
  }
}

addNavEvents();

(window as any).YTConfig = {
  host: "https://www.youtube.com"
};

interface LoginCookie {
  username: string;
}

const cookieStr = document.cookie;
if (cookieStr) {
  const loginUi: LoginCookie = JSON.parse(
    JSON.parse(cookieStr.split("; ")[0].split("=")[1])
  );
  const loginLink = document.getElementById("login");
  if (loginLink) {
    loginLink.innerText = loginUi.username;
    loginLink.setAttribute("href", "/logout");
  }
}
