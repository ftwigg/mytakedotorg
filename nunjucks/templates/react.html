{% extends "reactCore.html" %}

{% block bundle %}
	<!-- Main -->
	<script src="/app.bundle.js"></script>
{% endblock %}
