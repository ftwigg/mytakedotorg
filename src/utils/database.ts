import { FoundationTextType } from "../components/Foundation";
import { TakeBlock } from "../components/BlockEditor";

/**
 * Constraints:
 *  - titleSlug not null
 */
export interface Article {
  title: string;
  titleSlug: string;
  blocks: TakeBlock[];
}

/**
 * Constraints:
 *  - Username must be unique
 *  - All articles must have unique titleSlugs
 */
export interface User {
  name: string;
  articles: Article[] | null;
}

/**
 * Constraints:
 *  - Filename must be unique
 */
export interface DocumentExcerpt {
  title: string;
  filename: string;
  primaryDate: Date;
  primaryDateKind: "ratified" | "published";
}

/**
 * Constraints:
 *  - ID must be unique
 */
export interface Video {
  id: string;
  title: string;
  primaryDate: Date;
  primaryDateKind: "recorded";
}

export function isDocument(
  fact: DocumentExcerpt | Video | null
): fact is DocumentExcerpt {
  if (fact) {
    return (fact as DocumentExcerpt).filename !== undefined;
  } else {
    return false;
  }
}

export function isVideo(fact: DocumentExcerpt | Video | null): fact is Video {
  if (fact) {
    return (fact as Video).id !== undefined;
  } else {
    return false;
  }
}

export interface Database {
  excerpts: DocumentExcerpt[];
  videos: Video[];
  users: User[];
}

const database: Database = {
  videos: [
    {
      id: "qAqIKybNO38",
      title: "Jimmy Carter - Gerald Ford #1",
      primaryDate: new Date("1976-09-23"),
      primaryDateKind: "recorded"
    },
    {
      id: "TjHjU0Eu26Y",
      title: "Jimmy Carter - Gerald Ford #2",
      primaryDate: new Date("1976-10-06"),
      primaryDateKind: "recorded"
    },
    {
      id: "CipT04S0bVE",
      title: "Jimmy Carter - Gerald Ford #3",
      primaryDate: new Date("1976-10-22"),
      primaryDateKind: "recorded"
    }
  ],
  excerpts: [
    {
      title: "United States Constitution",
      filename: "united-states-constitution.foundation.html",
      primaryDate: new Date("1788-06-21"),
      primaryDateKind: "ratified"
    },
    {
      title: "Bill of Rights",
      filename: "bill-of-rights.foundation.html",
      primaryDate: new Date("1791-12-15"),
      primaryDateKind: "ratified"
    },
    {
      title: "Amendment 11",
      filename: "amendment-11.foundation.html",
      primaryDate: new Date("1795-02-07"),
      primaryDateKind: "ratified"
    },
    {
      title: "Amendment 12",
      filename: "amendment-12.foundation.html",
      primaryDate: new Date("1804-06-15"),
      primaryDateKind: "ratified"
    },
    {
      title: "Amendment 13",
      filename: "amendment-13.foundation.html",
      primaryDate: new Date("1865-12-06"),
      primaryDateKind: "ratified"
    },
    {
      title: "Amendment 14",
      filename: "amendment-14.foundation.html",
      primaryDate: new Date("1868-07-09"),
      primaryDateKind: "ratified"
    },
    {
      title: "Amendment 15",
      filename: "amendment-15.foundation.html",
      primaryDate: new Date("1870-02-03"),
      primaryDateKind: "ratified"
    },
    {
      title: "Amendment 16",
      filename: "amendment-16.foundation.html",
      primaryDate: new Date("1913-02-03"),
      primaryDateKind: "ratified"
    },
    {
      title: "Amendment 17",
      filename: "amendment-17.foundation.html",
      primaryDate: new Date("1913-04-08"),
      primaryDateKind: "ratified"
    },
    {
      title: "Amendment 18",
      filename: "amendment-18.foundation.html",
      primaryDate: new Date("1919-01-16"),
      primaryDateKind: "ratified"
    },
    {
      title: "Amendment 19",
      filename: "amendment-19.foundation.html",
      primaryDate: new Date("1920-08-18"),
      primaryDateKind: "ratified"
    },
    {
      title: "Amendment 20",
      filename: "amendment-20.foundation.html",
      primaryDate: new Date("1933-01-23"),
      primaryDateKind: "ratified"
    },
    {
      title: "Amendment 21",
      filename: "amendment-21.foundation.html",
      primaryDate: new Date("1933-12-05"),
      primaryDateKind: "ratified"
    },
    {
      title: "Amendment 22",
      filename: "amendment-22.foundation.html",
      primaryDate: new Date("1951-02-27"),
      primaryDateKind: "ratified"
    },
    {
      title: "Amendment 23",
      filename: "amendment-23.foundation.html",
      primaryDate: new Date("1961-03-29"),
      primaryDateKind: "ratified"
    },
    {
      title: "Amendment 24",
      filename: "amendment-24.foundation.html",
      primaryDate: new Date("1964-01-23"),
      primaryDateKind: "ratified"
    },
    {
      title: "Amendment 25",
      filename: "amendment-25.foundation.html",
      primaryDate: new Date("1967-02-10"),
      primaryDateKind: "ratified"
    },
    {
      title: "Amendment 26",
      filename: "amendment-26.foundation.html",
      primaryDate: new Date("1971-07-01"),
      primaryDateKind: "ratified"
    },
    {
      title: "Amendment 27",
      filename: "amendment-27.foundation.html",
      primaryDate: new Date("1992-05-05"),
      primaryDateKind: "ratified"
    }
  ],
  users: [
    {
      name: "samples",
      articles: [
        {
          title: "Does a law mean what it says, or what it meant?",
          titleSlug: "does-a-law-mean-what-it-says-or-what-it-meant",
          blocks: [
            {
              kind: "paragraph",
              text:
                "King: “I hereby declare that Seabiscuit shall be my royal horse!”"
            },
            {
              kind: "paragraph",
              text: "Subject: “But that’s MY horse!!”"
            },
            {
              kind: "paragraph",
              text:
                "King: “I said it – it’s the law of the land. Thief! Return my horse to me!”"
            },
            {
              kind: "paragraph",
              text: "..."
            },
            {
              kind: "paragraph",
              text:
                "That’s how people lived for a long time. The fix was to write laws down, so that the power of the government would no longer be subject to the arbitrary whims of any one person. Only trouble is, what does a law mean?"
            },
            {
              kind: "paragraph",
              text:
                "Textualism says that a law means what it says in its text. Simple enough. On the other hand, Originalism says that a law means what was meant when it was originally written. Figuring this out is a little trickier – you’ve got to be not just a lawyer but also an historian to have any chance of understanding the law. The upside of Originalism is that it can help the law to be more stable against the whims of a changing culture – our Constitution is more than 220 years old, and a lot has changed since then! A great example is the 2nd amendment:"
            },
            {
              kind: "document",
              excerptId: "bill-of-rights",
              highlightedRange: [368, 513]
            },
            {
              kind: "paragraph",
              text:
                "The Arms of the 21st century are a lot different than the Arms of 1789! A textualist has little choice in interpreting this law – clearly automatic rifles, grenades, rockets, and even nukes are protected. They are Arms, afterall!"
            },
            {
              kind: "paragraph",
              text:
                "An originalist has to go back to 1789 to understand the original meaning. Black powder muskets were the most powerful Arms of the time, so perhaps that is what the law should protect today? Although that captures the original technology, it doesn’t capture the original intent."
            },
            {
              kind: "document",
              excerptId: "bill-of-rights",
              highlightedRange: [368, 442]
            },
            {
              kind: "paragraph",
              text:
                "In 1776, George Washington led an American militia to victory against a tyrannical royal crown. As the newly free colonies organized their national government, many were afraid that it might be just as oppressive as the royal crown had been. It sure seems like the original intent of the 2nd amendment was to enshrine the right of the citizenry to have the means for armed rebellion."
            },
            {
              kind: "paragraph",
              text:
                "If our judges get to decide that the law means whatever they want it to mean, then we still live under a King, we’ve just changed his name to Judge. That’s why it’s important to apply consistent methodology when interpreting the law. But in this case, whether you’re a Textualist or an Originalist, I don’t see how the 2nd amendment allows for any of the gun restrictions in place today."
            }
          ]
        },
        {
          title: "Don't worry, we'll protect the Constitution for you!",
          titleSlug: "dont-worry-well-protect-the-constitution-for-you",
          blocks: [
            {
              kind: "paragraph",
              text:
                "The system for appointing judges to the Supreme Court is quite complex. Only a lawyer with years of experience could understand it. But trust me – this President has no right to make an appointment! We’ll let the people decide in the next election, as demanded by our Constitution!!"
            },
            {
              kind: "document",
              excerptId: "united-states-constitution",
              highlightedRange: [19343, 19970]
            },
            {
              kind: "paragraph",
              text:
                "Don’t bother reading the Constitution, it’s only for smart people like me. Trust me – the President is way out of line to think the Senate has any obligation to give his candidate a hearing!"
            }
          ]
        }
      ]
    }
  ]
};

export default database;
