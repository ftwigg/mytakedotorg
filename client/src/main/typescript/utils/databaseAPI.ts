import database, {
  CaptionMeta,
  CaptionWord,
  Database,
  VideoFact,
  User,
  Article
} from "./databaseData";
import { Foundation } from "../java2ts/Foundation";
import { Routes } from "../java2ts/Routes";
import { slugify } from "./functions";
import { TakeBlock } from "../components/BlockEditor";

export function getAllVideoFacts(): VideoFact[] {
  return database.videos;
}

export function getVideoFact(videoId: string): VideoFact {
  for (let video of database.videos) {
    if (video.id === videoId) {
      return video;
    }
  }

  throw "Cannot get VideoFact for video with id: " + videoId;
}

export function getVideoFactTitle(videoId: string): string {
  let video = getVideoFact(videoId);
  if (video) {
    return video.title;
  }

  throw "Cannot get video title for video with id: " + videoId;
}

export function getVideoFactPrimaryDate(videoId: string): Date {
  let video = getVideoFact(videoId);
  if (video) {
    return video.primaryDate;
  }

  throw "Cannot get video primary date for video with id: " + videoId;
}

export function getVideoFactPrimaryDateKind(videoId: string): string {
  let video = getVideoFact(videoId);
  if (video) {
    return video.primaryDateKind;
  }

  throw "Cannot get video primary date kind for video with id: " + videoId;
}

export function getVideoFactCaptionFile(videoId: string): string {
  let video = getVideoFact(videoId);
  if (video && video.captionFile) {
    return video.captionFile;
  }

  throw "Cannot get video caption file for video with id: " + videoId;
}

export function getVideoCaptionMetaData(videoId: string): CaptionMeta {
  let video = getVideoFact(videoId);
  if (video && video.captionMeta) {
    return video.captionMeta;
  }

  throw "Cannot get video caption metadata for video with id: " + videoId;
}

export function getVideoCaptionWordMap(videoId: string): CaptionWord[] {
  const captionFile = getVideoFactCaptionFile(videoId);
  const source = require("../foundation/" + captionFile);
  if (source) {
    return <CaptionWord[]>JSON.parse(source);
  }

  throw "Cannot get video caption word-map for video with id: " + videoId;
}

export function getAllFacts(
  callback: (
    error: string | Error | null,
    documents: Foundation.FactLink[]
  ) => any
): void {
  const headers = new Headers();

  headers.append("Accept", "application/json"); // This one is enough for GET requests

  const request: RequestInit = {
    method: "GET",
    headers: headers,
    cache: "default"
  };

  fetch(Routes.FOUNDATION_DATA_INDEX, request)
    .then(function(response: Response) {
      const contentType = response.headers.get("content-type");
      if (
        contentType &&
        contentType.indexOf("application/json") >= 0 &&
        response.ok
      ) {
        return response.json();
      } else {
        callback("Error retrieving Document Facts", []);
      }
    })
    .then(function(json: any) {
      callback(null, json);
    })
    .catch(function(error: Error) {
      // Network error
      callback(error, []);
    });
}

export function getDocumentFact(
  factHash: string,
  callback: (
    error: string | Error | null,
    documentFact: Foundation.DocumentFactContent | null
  ) => any
): void {
  const headers = new Headers();

  headers.append("Accept", "application/json"); // This one is enough for GET requests

  const request: RequestInit = {
    method: "GET",
    headers: headers,
    cache: "default"
  };

  fetch(Routes.FOUNDATION_DATA + "/" + factHash + ".json", request)
    .then(function(response: Response) {
      const contentType = response.headers.get("content-type");
      if (
        contentType &&
        contentType.indexOf("application/json") >= 0 &&
        response.ok
      ) {
        return response.json();
      } else {
        callback("Error retrieving Document Fact", null);
      }
    })
    .then(function(json: any) {
      callback(null, json);
    })
    .catch(function(error: Error) {
      // Network error
      callback(error, null);
    });
}

export function getDocumentFactTitle(excerptId: string): string {
  for (let excerpt of database.excerpts) {
    if (slugify(excerpt.title) === excerptId) {
      return excerpt.title;
    }
  }

  throw "Cannot get document title for document with id: " + excerptId;
}

export function getDocumentFactFilename(excerptId: string): string {
  for (let excerpt of database.excerpts) {
    if (slugify(excerpt.title) === excerptId) {
      return excerpt.filename;
    }
  }

  throw "Cannot get document filename for document with id: " + excerptId;
}

export function getDocumentFactPrimaryDate(excerptId: string): Date {
  for (let excerpt of database.excerpts) {
    if (slugify(excerpt.title) === excerptId) {
      return excerpt.primaryDate;
    }
  }

  throw "Cannot get document primary date for document with id: " + excerptId;
}

export function getDocumentFactPrimaryDateKind(excerptId: string): string {
  for (let excerpt of database.excerpts) {
    if (slugify(excerpt.title) === excerptId) {
      return excerpt.primaryDateKind;
    }
  }

  throw "Cannot get document primary date kind for document with id: " +
    excerptId;
}

export function getUser(username: string): User | null {
  for (let user of database.users) {
    if (user.name === username) {
      return user;
    }
  }

  throw "Cannot get user: " + username;
}

export function getAllUsers(): User[] {
  return database.users;
}

export function createUser(username: string): boolean {
  if (getUser(username)) {
    throw "User already exists with username: " + username;
  }

  return database.users.push({ name: username, articles: null }) > 0;
}

export function updateUserName(
  oldUserName: string,
  newUserName: string
): boolean {
  let oldUser = getUser(oldUserName);
  let newUser = getUser(newUserName);
  if (!oldUser || newUser) {
    throw "Failed to update username";
  }

  oldUser.name = newUserName;

  return true;
}

export function deleteUser(username: string): boolean {
  if (!getUser(username)) {
    throw "User doesn't exist: " + username;
  }

  let newDatabase: Database = {
    ...database,
    users: database.users.filter(user => {
      return user.name !== username;
    })
  };

  return true;
}

export function getAllArticles(username?: string): Article[] | null {
  if (username) {
    let user = getUser(username);
    if (user) {
      return user.articles;
    } else {
      return null;
    }
  }

  let articles: Article[] = [];
  for (let user of database.users) {
    if (user.articles) {
      articles = [...articles, ...user.articles];
    }
  }

  return articles;
}

export function getArticle(
  username: string,
  titleSlug: string
): Article | null {
  let user = getUser(username);
  if (user && user.articles) {
    for (let article of user.articles) {
      if (article.titleSlug === titleSlug) {
        return article;
      }
    }
  } else {
    //Invalid username
    throw "User doesn't exist: " + username;
  }

  //Article not found
  return null;
}

export function createArticle(
  username: string,
  title: string,
  blocks: TakeBlock[],
  previewBlocks: number[]
): boolean {
  if (!title) {
    // Throw error, title is required
    throw "Invalid title: " + title;
  }
  if (getArticle(username, slugify(title))) {
    // Can't create duplicate article
    throw "User, " + username + ", already has article with title, " + title;
    /**
		 * Could also create the article anyway and append an ID counter to the title,
		 * e.g. new-take-title, new-take-title-2, new-take-title-3
		 */
  }
  let user = getUser(username);
  if (user) {
    let articles = getAllArticles(username);
    if (articles) {
      articles = [
        ...articles,
        {
          title: title,
          titleSlug: slugify(title),
          blocks: blocks,
          previewBlocks: previewBlocks
        }
      ];
    } else {
      articles = [
        {
          title: title,
          titleSlug: slugify(title),
          blocks: blocks,
          previewBlocks: previewBlocks
        }
      ];
    }
    return true;
  } else {
    // Invalid username
    throw "User doesn't exist: " + username;
  }
}

export function updateArticleTitle(
  username: string,
  oldTitleSlug: string,
  newTitle: string
): boolean {
  let newArticle = getArticle(username, slugify(newTitle));
  if (newArticle) {
    // Can't create duplicate article
    throw "User, " + username + ", already has article with title, " + newTitle;
    /**
		 * Could also create the article anyway and append an ID counter to the title,
		 * e.g. new-take-title, new-take-title-2, new-take-title-3
		 */
  }

  let oldArticle = getArticle(username, oldTitleSlug);
  if (oldArticle) {
    oldArticle.title = newTitle;
    oldArticle.titleSlug = slugify(newTitle);

    return true;
  }

  throw "Failed to update article title";
}

export function updateArticle(
  username: string,
  titleSlug: string,
  newBlocks: TakeBlock[]
): boolean {
  let article = getArticle(username, titleSlug);
  if (article) {
    article.blocks = newBlocks;

    return true;
  }

  throw "Failed to update article";
}

export function deleteArticle(username: string, titleSlug: string): boolean {
  let user = getUser(username);
  if (user && user.articles) {
    let articles = user.articles.filter(article => {
      return article.titleSlug !== titleSlug;
    });
    let newUser: User = {
      ...user,
      articles: articles
    };
    let newDatabase: Database = {
      ...database,
      users: [...database.users, newUser]
    };
    return true;
  } else {
    // User or articles don't exist
    throw "Failed to delete article";
  }
}
