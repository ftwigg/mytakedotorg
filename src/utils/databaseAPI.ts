import database, {
  Database,
  VideoFact,
  DocumentFact,
  User,
  Article
} from "./databaseData";
import { slugify } from "./functions";
import { TakeBlock } from "../components/BlockEditor";

export function getAllVideoFacts(): VideoFact[] {
  return database.videos;
}

export function getVideoFact(videoId: string): VideoFact | null {
  for (let video of database.videos) {
    if (video.id === videoId) {
      return video;
    }
  }

  return null;
}

export function getVideoFactTitle(videoId: string): string | null {
  let video = getVideoFact(videoId);
  if (video) {
    return video.title;
  }

  return null;
}

export function getVideoFactPrimaryDate(videoId: string): Date | null {
  let video = getVideoFact(videoId);
  if (video) {
    return video.primaryDate;
  }

  return null;
}

export function getVideoFactPrimaryDateKind(videoId: string): string | null {
  let video = getVideoFact(videoId);
  if (video) {
    return video.primaryDateKind;
  }

  return null;
}

export function getVideoFactCaptionFile(videoId: string): string | null {
  let video = getVideoFact(videoId);
  if (video && video.captionFile) {
    return video.captionFile;
  }

  return null;
}

export function getAllDocumentFacts(): DocumentFact[] {
  return database.excerpts;
}

export function getDocumentFact(excerptId: string): DocumentFact | null {
  for (let excerpt of database.excerpts) {
    if (slugify(excerpt.title) === excerptId) {
      return excerpt;
    }
  }

  return null;
}

export function getDocumentFactTitle(excerptId: string): string | null {
  for (let excerpt of database.excerpts) {
    if (slugify(excerpt.title) === excerptId) {
      return excerpt.title;
    }
  }

  return null;
}

export function getDocumentFactFilename(excerptId: string): string | null {
  for (let excerpt of database.excerpts) {
    if (slugify(excerpt.title) === excerptId) {
      return excerpt.filename;
    }
  }

  return null;
}

export function getDocumentFactPrimaryDate(excerptId: string): Date | null {
  for (let excerpt of database.excerpts) {
    if (slugify(excerpt.title) === excerptId) {
      return excerpt.primaryDate;
    }
  }

  return null;
}

export function getDocumentFactPrimaryDateKind(
  excerptId: string
): string | null {
  for (let excerpt of database.excerpts) {
    if (slugify(excerpt.title) === excerptId) {
      return excerpt.primaryDateKind;
    }
  }

  return null;
}

export function getUser(username: string): User | null {
  for (let user of database.users) {
    if (user.name === username) {
      return user;
    }
  }

  return null;
}

export function getAllUsers(): User[] {
  return database.users;
}

export function createUser(username: string): boolean {
  if (getUser(username)) {
    return false;
  }

  return database.users.push({ name: username, articles: null }) > 0;
}

export function updateUserName(
  oldUserName: string,
  newUserName: string
): boolean {
  let oldUser = getUser(oldUserName);
  let newUser = getUser(newUserName);
  if (!oldUser || newUser) {
    return false;
  }

  oldUser.name = newUserName;

  return true;
}

export function deleteUser(username: string): boolean {
  if (!getUser(username)) {
    // User doesn't exist
    return false;
  }

  let newDatabase: Database = {
    ...database,
    users: database.users.filter(user => {
      return user.name !== username;
    })
  };

  return true;
}

export function getAllArticles(username?: string): Article[] | null {
  if (username) {
    let user = getUser(username);
    if (user) {
      return user.articles;
    } else {
      return null;
    }
  }

  let articles: Article[] = [];
  for (let user of database.users) {
    if (user.articles) {
      articles = [...articles, ...user.articles];
    }
  }

  return articles;
}

export function getArticle(
  username: string,
  titleSlug: string
): Article | null {
  let user = getUser(username);
  if (user && user.articles) {
    for (let article of user.articles) {
      if (article.titleSlug === titleSlug) {
        return article;
      }
    }
  } else {
    //Invalid username
    return null;
  }

  //Article not found
  return null;
}

export function createArticle(
  username: string,
  title: string,
  blocks: TakeBlock[],
  previewBlocks: number[]
): boolean {
  if (!title) {
    // Throw error, title is required
    return false;
  }
  if (getArticle(username, slugify(title))) {
    // Can't create duplicate article
    return false;
    /**
		 * Could also create the article anyway and append an ID counter to the title,
		 * e.g. new-take-title, new-take-title-2, new-take-title-3
		 */
  }
  let user = getUser(username);
  if (user) {
    let articles = getAllArticles(username);
    if (articles) {
      articles = [
        ...articles,
        {
          title: title,
          titleSlug: slugify(title),
          blocks: blocks,
          previewBlocks: previewBlocks
        }
      ];
    } else {
      articles = [
        {
          title: title,
          titleSlug: slugify(title),
          blocks: blocks,
          previewBlocks: previewBlocks
        }
      ];
    }
    return true;
  } else {
    // Invalid username
    return false;
  }
}

export function updateArticleTitle(
  username: string,
  oldTitleSlug: string,
  newTitle: string
): boolean {
  let newArticle = getArticle(username, slugify(newTitle));
  if (newArticle) {
    // Can't create duplicate article
    return false;
    /**
		 * Could also create the article anyway and append an ID counter to the title,
		 * e.g. new-take-title, new-take-title-2, new-take-title-3
		 */
  }

  let oldArticle = getArticle(username, oldTitleSlug);
  if (oldArticle) {
    oldArticle.title = newTitle;
    oldArticle.titleSlug = slugify(newTitle);

    return true;
  }

  return false;
}

export function updateArticle(
  username: string,
  titleSlug: string,
  newBlocks: TakeBlock[]
): boolean {
  let article = getArticle(username, titleSlug);
  if (article) {
    article.blocks = newBlocks;

    return true;
  }

  return false;
}

export function deleteArticle(username: string, titleSlug: string): boolean {
  let user = getUser(username);
  if (user && user.articles) {
    let articles = user.articles.filter(article => {
      return article.titleSlug !== titleSlug;
    });
    let newUser: User = {
      ...user,
      articles: articles
    };
    let newDatabase: Database = {
      ...database,
      users: [...database.users, newUser]
    };
    return true;
  } else {
    // User or articles don't exist
    return false;
  }
}
