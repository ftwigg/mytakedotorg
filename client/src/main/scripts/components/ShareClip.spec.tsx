import * as React from "react";
import * as renderer from "react-test-renderer";
import ShareClip from "./ShareClip";
import { videoFactLink } from "../utils/testUtils";

test("It renders", () => {
  const tree = renderer
    .create(
      <ShareClip
        factSlug={"presidential-debate-clinton-trump-2-of-3"}
        highlightedRange={[0.1, 5]}
        videoIdHash={videoFactLink.hash}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
