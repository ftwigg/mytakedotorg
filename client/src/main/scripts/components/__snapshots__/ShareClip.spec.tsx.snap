// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`It renders 1`] = `
<div
  className="shareclip"
>
  <h3
    className="shareclip__title"
  >
    Share your clip
  </h3>
  <div>
    <label
      className="shareclip__label"
      htmlFor="cliptitle"
    >
      Title
    </label>
    <input
      className="shareclip__input"
      id="cliptitle"
      onChange={[Function]}
      type="text"
      value=""
    />
    <pre
      className="shareclip__url"
    >
      <code>
        about:///anonymous/untitled/v1/eyJ0aXRsZSI6InVudGl0bGVkIiwidmlkSWQiOiJpRWZ3SXhNN01tbk9LYjd6dDRIcVc4SXhVV3k2RjdhMjM2ZlNPUWxVVVdJPSIsImhTdGFydCI6IjAuMTAwIiwiaEVuZCI6IjUuMDAwIn0=
      </code>
    </pre>
    <button
      className="shareclip__button"
      onClick={[Function]}
    >
      Copy link
    </button>
    <button
      className="shareclip__button"
      onClick={[Function]}
    >
      Facebook
    </button>
    <button
      className="shareclip__button"
      onClick={[Function]}
    >
      Twitter
    </button>
    <button
      className="shareclip__button"
      onClick={[Function]}
    >
      Give your Take
    </button>
    <div
      className="shareclip__info"
    />
  </div>
</div>
`;
